src/FunctionApps/python/GenerateCSVs/ecr.py

```diff
@@ -1,8 +1,96 @@
-from typing import List, Any
-from GenerateCSVs.patient import PATIENT_COLUMNS
+import logging
 
-ECR_COLUMNS = PATIENT_COLUMNS + []
+from typing import List
+from GenerateCSVs.patient import PATIENT_COLUMNS, parse_patient_resource
+from GenerateCSVs.elr import extract_loinc_lab
+from GenerateCSVs.vxu import parse_immunization_resource
 
+ECR_COLUMNS = (
+    PATIENT_COLUMNS
+    + [  # Immunization columns
+        "immunizationVaccineCode",
+        "immunizationVaccineDescription",
+        "immunizationOccurrenceDateTime",
+    ]
+    + [  # Observation columns
+        "observationLoincCode",
+        "observationResult",
+        "observationEffectiveDateTime",
+    ]
+)
 
-def ecr_to_csv(bundle: dict) -> List[Any]:
-    pass
+
+def ecr_to_csv(bundle: dict) -> List[List[str]]:
+    """
+    Given a FHIR bundle containing a patient resource and one or more
+    immunization resources, create a row containing data specified by
+    ECR_COLUMNS.
+    """
+    # Initialize variables
+    patient_resources = {}
+    entries = []
+    return_rows = []
+
+    # Loop through bundle, and build:
+    #  * A dictionary to index patient resources
+    #  * A list of immunization records that should appear in the CSV
+    #  * A list of observation records that should appear in the CSV
+    for entry in bundle.get("entry"):
+        resource = entry["resource"]
+        if resource.get("resourceType") == "Patient" and len(resource.get("id")) > 0:
+            patient_resources["Patient/" + resource.get("id")] = parse_patient_resource(
+                entry
+            )
+        elif resource.get("resourceType") in ["Immunization", "Observation"]:
+            entries.append(entry)
+
+    # Loop through immunization resources collected from bundle.
+    # For each immunization resource, pull in its corresponding patient
+    # and join immunization-specific fields.
+    for entry in entries:
```

Comment:
Is there a reason to store the immunization and observation entries and then loop through them again instead of processing them in the original loop? This approach results in looping through the results twice, which I believe can be avoided if you were to create three functions that processed the entry in the initial pass. Use `process_patient_entry`, `process_immunization_entry`, and `process_observation_entry` to pass any given valid entry, determined by an initial check of `resource.get("resourceType")` and append the results to the appropriate collections.